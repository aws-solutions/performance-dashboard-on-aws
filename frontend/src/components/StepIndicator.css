.usa-step-indicator {
  background-color: transparent !important;
}

.usa-step-indicator__segment--current::after {
  background-color: #565c65 !important;
}

.usa-step-indicator__segment--complete::after {
  background-color: #1b1b1b !important;
}

.usa-step-indicator--counters .usa-step-indicator__segment--current::before,
.usa-step-indicator--counters-sm .usa-step-indicator__segment--current::before {
  background-color: #565c65 !important;
}

.usa-step-indicator--counters .usa-step-indicator__segment--complete::before,
.usa-step-indicator--counters-sm .usa-step-indicator__segment--complete::before {
  background-color: #1b1b1b !important;
}
