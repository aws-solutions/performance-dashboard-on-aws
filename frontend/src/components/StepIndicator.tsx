/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "./StepIndicator.css";
import { useTranslation } from "react-i18next";

interface Props {
    current: number;
    segments: Array<{
        label: string;
    }>;
    showStepChart: boolean;
    showStepText: boolean;
}

function StepIndicator(props: Props) {
    const { t } = useTranslation();
    if (props.current >= props.segments.length || props.current < 0) {
        return null;
    }

    const stepChart = (
        <ol className="usa-step-indicator__segments" aria-label={t("Progress")}>
            {props.segments.map((segment, index) => {
                return (
                    <Segment
                        key={index}
                        current={props.current === index}
                        complete={index < props.current}
                        label={segment.label}
                    />
                );
            })}
        </ol>
    );

    const stepText = (
        <div className="usa-step-indicator__header">
            <h2 className="usa-step-indicator__heading">
                <span className="usa-step-indicator__heading-counter">
                    <span className="usa-sr-only">{t("Step")}</span>
                    <span className="usa-step-indicator__current-step bg-base-dark">
                        {props.current + 1}
                    </span>
                    <span className="usa-step-indicator__total-steps text-base-dark margin-left-2px">
                        {t("Of")} {props.segments.length}
                    </span>
                </span>
                <span className="usa-step-indicator__heading-text">
                    {props.current < props.segments.length && props.segments[props.current].label}
                </span>
            </h2>
        </div>
    );

    return (
        <div
            className="usa-step-indicator usa-step-indicator--counters-sm"
            role="progressbar"
            aria-valuemin={1}
            aria-valuemax={props.segments.length}
            aria-valuenow={props.current + 1}
            aria-valuetext={t("StepIndicatorLabel", {
                current: props.current + 1,
                total: props.segments.length,
                title: props.segments[props.current].label,
            })}
        >
            {props.showStepChart ? stepChart : ""}
            {props.showStepText ? stepText : ""}
        </div>
    );
}

interface SegmentProps {
    current?: boolean;
    complete?: boolean;
    label: string;
}

function Segment(props: SegmentProps) {
    const { t } = useTranslation();
    if (props.current) {
        return (
            <li
                className="usa-step-indicator__segment usa-step-indicator__segment--current"
                aria-current="step"
            >
                <span className="usa-step-indicator__segment-label">
                    <div className="text-base-dark">{props.label}</div>
                </span>
            </li>
        );
    }

    if (props.complete) {
        return (
            <li className="usa-step-indicator__segment usa-step-indicator__segment--complete">
                <span className="usa-step-indicator__segment-label">
                    <div className="text-base-dark">{props.label}</div>{" "}
                    <span className="usa-sr-only">{t("Completed")}</span>
                </span>
            </li>
        );
    }

    return (
        <li className="usa-step-indicator__segment">
            <span className="usa-step-indicator__segment-label">
                <div className="text-base-dark">{props.label}</div>{" "}
                <span className="usa-sr-only">{t("NotCompleted")}</span>
            </span>
        </li>
    );
}

export default StepIndicator;
