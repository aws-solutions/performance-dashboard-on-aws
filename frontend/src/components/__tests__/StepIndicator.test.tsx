/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import StepIndicator from "../StepIndicator";

test("renders a step indicator with multiple segments", async () => {
    const wrapper = render(
        <StepIndicator
            current={0}
            segments={[
                {
                    label: "Step 1",
                },
                {
                    label: "Step 2",
                },
                {
                    label: "Step 3",
                },
            ]}
            showStepChart={true}
            showStepText={false}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a step indicator with one completed step", async () => {
    const wrapper = render(
        <StepIndicator
            current={1}
            segments={[
                {
                    label: "Step 1",
                },
                {
                    label: "Step 2",
                },
                {
                    label: "Step 3",
                },
            ]}
            showStepChart={true}
            showStepText={false}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("returns null when current is out of bounds", async () => {
    const wrapper = render(
        <StepIndicator
            current={-1}
            segments={[
                {
                    label: "Step 1",
                },
            ]}
            showStepChart={true}
            showStepText={false}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});
