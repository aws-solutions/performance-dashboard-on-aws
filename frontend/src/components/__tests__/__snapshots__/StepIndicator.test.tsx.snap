// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a step indicator with multiple segments 1`] = `
<div>
  <div
    aria-valuemax="3"
    aria-valuemin="1"
    aria-valuenow="1"
    aria-valuetext="1 of 3: Step 1"
    class="usa-step-indicator usa-step-indicator--counters-sm"
    role="progressbar"
  >
    <ol
      aria-label="progress"
      class="usa-step-indicator__segments"
    >
      <li
        aria-current="step"
        class="usa-step-indicator__segment usa-step-indicator__segment--current"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 1
          </div>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 2
          </div>
           
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 3
          </div>
           
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
    </ol>
    
  </div>
</div>
`;

exports[`renders a step indicator with one completed step 1`] = `
<div>
  <div
    aria-valuemax="3"
    aria-valuemin="1"
    aria-valuenow="2"
    aria-valuetext="2 of 3: Step 2"
    class="usa-step-indicator usa-step-indicator--counters-sm"
    role="progressbar"
  >
    <ol
      aria-label="progress"
      class="usa-step-indicator__segments"
    >
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--complete"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 1
          </div>
           
          <span
            class="usa-sr-only"
          >
            completed
          </span>
        </span>
      </li>
      <li
        aria-current="step"
        class="usa-step-indicator__segment usa-step-indicator__segment--current"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 2
          </div>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          <div
            class="text-base-dark"
          >
            Step 3
          </div>
           
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
    </ol>
    
  </div>
</div>
`;

exports[`returns null when current is out of bounds 1`] = `<div />`;
